### ISSUE TITLE
Quantized Mobilenet using ARMNN gives wrong results

### ISSUE THREAD

**Issue by uu-praveeng** [ARM-software/armnn#344]
Hi,

I have created a mobilenetv2 FP32 armnn model using ARMNN APIs. ARMNN FP32 model inference  is as expected and classified output correctly for bell pepper image.

When I quantized the same network using ArmnnQuantizer for bell pepper image and run inference using INT8 model we are seeing completely different results.
Could you look into this issue.

I have attached the below files for you reference [MobileNetV2.zip](https://github.com/ARM-software/armnn/files/4218928/MobileNetV2.zip)
1. Mobilenet.cpp - for FP32 inference and armnn model creation
2. img.txt for FP32 inference 
3. img.raw , csv file used for quantization
4. Quantized and FP32 models
5. MobilenetINT8.cpp used for INT8 inference.

Let me know if you need any further info.
Note: The same exercise with ResNet model were giving convincing results for FP32 vs armnn.

Thank you,
Praveen.

**Comment by FrancisMurtagh-arm** [comment]
Hi,

I'm looking into this now.

Thanks,
Francis.

**Comment by FrancisMurtagh-arm** [comment]
Hi uu-praveeng,

When I run Mobilenet.cpp with img.txt as input my output is 0.001 for all classes.

Could you attach your output.txt please?

Thanks,
Francis.

**Comment by uu-praveeng** [comment]
Hi @FrancisMurtagh-arm 
It is expected because I have not supplied the weights,biases .It will be taking some garbage values for weights/biases  and result could be what you are seeing.
MobileNet.cpp was supplied to give idea what are the layers present in the network.

Could you try to use the MobileNet.armnn which is derived from MobileNet.cpp, and run the inference you should be seeing 0.9722 for 946 index.

Thank you,
Praveen.

**Comment by FrancisMurtagh-arm** [comment]
Hi uu-praveeng,

Thanks for the info, I used our ExecuteNetwork for inference and got 0.878826 for index 946.
I'll try quantizing and see if how it goes.

Francis.

**Comment by FrancisMurtagh-arm** [comment]
Hi uu-praveeng,

Did you use img.raw as it is for input to ArmnnQuantizer?

Using it as follows gives an error for me.
./ArmnnQuantizer -c ./img.raw -s QAsymm8 -f ./MobileNet.armnn -d . -o quantMobileNet.armnn

terminate called after throwing an instance of 'armnn::Exception'
  what():  CSV file img.raw] does not have correct number of entries on line 0. Expected 3 entries but was 1
Aborted (core dumped)

Thanks,
Francis.

**Comment by uu-praveeng** [comment]
Hi @FrancisMurtagh-arm ,
You need to use the below command to quantize.

./ArmnnQuantizer -c ./MobileNet.csv -s QAsymm8 -f ./MobileNet.armnn -d . -o quantMobileNet.armnn

Thank you,
Praveen.

**Comment by FrancisMurtagh-arm** [comment]
Hi  uu-praveeng,

Quantizing as above and running inferences now gives all zeroes for classes. Is your img.raw just an example or are you just quantizing with one image tensor?

Thanks,
Francis.

**Comment by uu-praveeng** [comment]
Hi @FrancisMurtagh-arm 

img.raw is a derived from img.txt. We are trying to quantize the armnn FP32 model with single image and then run the inference with the same img.txt . 
This essentially makes sure that Quantized model is fool proof and as a next step we could quantized it for a bigger dataset .

Since quantizing using same image and int8 inference with that image is not as expected I am unclear if there are any issues with quantization module itself or Does it have to do with Quantization will only work for certain models and not for other.

Note: I have followed the same process for ResNet and int8 inference was as expected for img.txt
Thank you,
Praveen.

**Comment by FrancisMurtagh-arm** [comment]
Hi Praveen,

`./ArmnnQuantizer -c ./MobileNet.csv -s QAsymm8 -f ./MobileNet.armnn -d . -o quantMobileNet.armnn`

Could you try to quantize with the above command and run inference on the resulting model to see if you get the same result?

I'm not able to reproduce your results for the quantized model; it is still giving zeroes for inference:


[issue344_output.txt](https://github.com/ARM-software/armnn/files/4268173/issue344_output.txt)
[quantMobileNet.armnn.zip](https://github.com/ARM-software/armnn/files/4268176/quantMobileNet.armnn.zip)

Kind Regards,
Francis.

**Comment by uu-praveeng** [comment]
Hi Francis,

Please use this command "./ArmnnQuantizer -f MobileNet.armnn -c MobileNet.csv -p -d . -o QuantMobileNet.armnn" for given armnn model

Once you have the quantized model .Use the command                                                                        " ./MobileNetINT8Exe QuantMobileNet.armnn img.txt INT8Out.txt"  to get QuantOut.txt. I have attached the INT8Out.txt for you reference and FP32Out.txt for your reference.
(MobileNetINT8Exe is derived by compiling the MobileNetINT8Txt.cpp file that I have shared)

[FP32Out.txt](https://github.com/ARM-software/armnn/files/4275506/FP32Out.txt)
[INT8Out.txt](https://github.com/ARM-software/armnn/files/4275508/INT8Out.txt)

Let me know if you need any more info.
Thank you,
Praveen.

**Comment by FrancisMurtagh-arm** [comment]
Hi Praveen,

I'm able to reproduce now when I add the preserve-data-type (-p) argument.

Much appreciated,
Francis.

**Comment by FrancisMurtagh-arm** [comment]
Hi Praveen,

Would you be able to provide the ResNet version of the model, which you had success with?

Thanks,
Francis.

**Comment by uu-praveeng** [comment]
Hi Francis,

I have attached the FP32 model- ResNet.armnn UINT8 model - QuantResNet.armnn and  the corresponding out.txt files.

https://drive.google.com/open?id=1r0sIstp2MvbOxwZ-AFLhW17v2jT6fg5U

Thank you,
Praveen.

**Comment by FrancisMurtagh-arm** [comment]
Hi Praveen,

Much appreciated,
Francis.

**Comment by FrancisMurtagh-arm** [comment]
Hi Praveen,

Do you have a similar Tensorflow or TfLlite format model that you could share?

Regards,
Francis.

**Comment by uu-praveeng** [comment]
Hi @FrancisMurtagh-arm 

Sorry, I do not have any reference tensorflow/tflite model. I have created both from scratch using armnn APIs.

Thank you,
Praveen.

**Comment by FrancisMurtagh-arm** [comment]
Hi Praveen,

I did some research into Post-Training Quantization, our current implementation performs quantization of weights at a per-layer granularity which is fine for ResNet as your exercise showed.

However MobileNetV1 and V2 are much smaller with fewer parameters resulting in vastly poorer accuracy results once quantized.

Google have done a Whitepaper on Quantizer design which covers your exact situation:
https://arxiv.org/pdf/1806.08342.pdf

In the table 3 of the paper you can see the results from their experiments after per-layer quantization the accuracies of  MobileNetV1 and MobileNetv2 drop to 0.001 and 0.004 respectively.

![PostTrainingQuantization](https://user-images.githubusercontent.com/44639508/76615590-40dcf900-651a-11ea-9a6d-3388625b1333.png)

They also observe:
> 3. Networks with more parameters like Resnets and Inception-v3 are more robust
> to quantization compared to Mobilenets which have fewer parameters.
> 4. There is a large drop when weights are quantized at the granularity of a layer,
> particularly for Mobilenet architectures.

For our Quantizer to be effective for MobileNet we would first require implementation of per-channel weight quantization in our Quantizer and Serializer.

TfLite have already implemented these features if using their tooling is an option.

Kind Regards,
Francis.

**Comment by james-conroy-arm** [comment]
Hi @uu-praveeng ,

Did you look into using TfLite for per-channel weight quantization tooling?

Cheers,
James

**Comment by uu-praveeng** [comment]
Hi @james-conroy-arm 

No, I didn't look into it. Do you have any such tflite model quantized per channel. 
I can try it out and see if it gives better results.

Thank you,
Praveen.

**Comment by waynexhsieh** [comment]
Hi @uu-praveeng 

How do you generate the raw file ?

I download the mobilenet v1 1.0 224 fp32 tflite model from Google official release, and use ArmnnConverter convert to "mobilenet.armnn"

`./ImageTensorGenerator -i xxx.jpg -f tflite -o xxx.raw --new-width 224 --new-height 224 -l "NCHW"`
`./ImageCSVFileGenerator -i . -o img.csv `
`./ArmnnQuantizer -f mobilenet.armnn -c ./img.csv -p -d . -o quantize.armnn`

I got an error at ArmnnQuantizer 
`terminating with uncaught exception of type armnn::Exception: Could not retrieve tensor info for binding ID 0`

Do you have any suggestions?

**Comment by uu-praveeng** [comment]
Hi @waynexhsieh 

I have used the attahced .cpp file to generate the raw file ,csv file . Let me know if this helps.
You need to pass the image in .txt format with UINT8 datatype.

Thanks,
Praveen.
[CSVTensorGeneratorArg.zip](https://github.com/ARM-software/armnn/files/5129000/CSVTensorGeneratorArg.zip)

**Comment by waynexhsieh** [comment]
Hi @uu-praveeng ,

The attachced.cpp file can't solve it.
I found that the "layerBindingId" of my generated model is not 0, but is 44288. (I don't know why is 44288)
And I change the number in csv file, and quantize the model success.

Thanks,
Wayne

**Comment by FrancisMurtagh-arm** [comment]
Hi @waynexhsieh,

The ImageCSVFileGenerator was given the optional "Input layer binding Id" argument to workaround this issue: https://review.mlplatform.org/c/ml/armnn/+/1764/

Regards,
Francis.

**Comment by FrancisMurtagh-arm** [comment]
Closing as original issue was due to short-comings of per-layer quantization in small models like MobileNet, as described by https://github.com/ARM-software/armnn/issues/344#issuecomment-598670629

If you need any more help @waynexhsieh please open a new issue.

Thanks,
Francis.
